File: rotate_arr.py
"""
Time Complexity: O(n)
Space Complexity: O(1)
"""
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if not nums or not k: return
        if len(nums) == 1: return
        if k > len(nums):
            k = k - len(nums)
        nums = self.reverse(nums, 0, len(nums)-1)
        nums = self.reverse(nums, 0, k-1)
        nums = self.reverse(nums, k, len(nums)-1)
        
    def reverse(self, nums, left, right):
        
        while left < right:
            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right -= 1
            
        return nums

File: rotateArr.java
// O(n)
class Solution {
    public void rotate(int[] nums, int k) {
        int n= nums.length;
        if(k>=n) k= k%n;             //
        System.out.println(k);
        reverse(nums,0,n-1);
        reverse(nums,0,k-1);
        reverse(nums,k, n-1);
    }
    
    private void reverse(int []nums, int start, int end){
        int i=start;
        int j=end;
        while(i<j){
            int temp= nums[i];
            nums[i]= nums[j];
            nums[j]= temp;
            i++;
            j--;
        }
    }
}
